File: 211. Design Add and Search Words Data Structure/sol.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"reflect"
)

type TrieNode struct {
	children  map[rune]*TrieNode
	endOfWord bool
}

func ConstructTrie() *TrieNode {
	return &TrieNode{children: make(map[rune]*TrieNode)}
}

type WordDictionary struct {
	root *TrieNode
}

func Constructor() WordDictionary {
	return WordDictionary{root: ConstructTrie()}
}

func (w *WordDictionary) addWord(word string) {
	curr := w.root
	for _, c := range word {
		if _, ok := curr.children[c]; !ok {
			curr.children[c] = ConstructTrie()
		}
		curr = curr.children[c]
	}
	curr.endOfWord = true

}

func (w *WordDictionary) search(word string) bool {

	var subSearch func(j int, curr *TrieNode) bool

	subSearch = func(j int, curr *TrieNode) bool {
		fmt.Println(word[j:], curr)
		for i := j; i < len(word); i++ {
			c := rune(word[i])
			fmt.Println(i, string(c))
			if c == '.' {
				for _, val := range curr.children {
					if subSearch(i+1, val) {
						return true
					}
				}
				return false

			}
			if _, ok := curr.children[c]; !ok {
				return false
			}
			curr = curr.children[c]
		}
		return curr.endOfWord
	}
	return subSearch(0, w.root)

}

func main() {
	file := "211. Design Add and Search Words Data Structure/testcase.json"
	path := filepath.Join(file)
	jsonData, err := os.ReadFile(path)
	if err != nil {
		fmt.Printf("Error reading json data : %v", err)
		return
	}
	var testcase TestCase
	err = json.Unmarshal(jsonData, &testcase)
	if err != nil {
		fmt.Printf("Error unmarshalling json data : %v", err)
		return
	}

	passNum := 0
	var wordDictionary WordDictionary

	for _, test := range testcase.Tests {

		output := []any{}
		for i := 0; i < len(test.Input.Operations); i++ {
			res := any(nil)
			switch test.Input.Operations[i] {
			case "WordDictionary":
				wordDictionary = Constructor()
			case "addWord":
				wordDictionary.addWord(test.Input.Arguments[i][0])
			case "search":
				res = wordDictionary.search(test.Input.Arguments[i][0])

			}
			output = append(output, res)

		}
		if reflect.DeepEqual(output, test.ExpectedOutput) {
			passNum++
			fmt.Println("Test passed")
		} else {
			fmt.Println("Test failed")
		}
		fmt.Printf("operations        : %v\n", test.Input.Operations)
		fmt.Printf("arguments         : %v\n", test.Input.Arguments)
		fmt.Printf("output            : %v\n", output)
		fmt.Printf("expected ouput    : %v\n", test.ExpectedOutput)
		fmt.Println("---------------")
	}
	fmt.Printf("Total tests : %v, Passed: %v, Failed: %v", len(testcase.Tests), passNum, len(testcase.Tests)-passNum)
}

type TestCase struct {
	Tests []Test `json:"tests"`
}
type Test struct {
	Input          Input `json:"input"`
	ExpectedOutput []any `json:"expectedOutput"`
}
type Input struct {
	Operations []string   `json:"operations"`
	Arguments  [][]string `json:"arguments"`
}

File: 211. Design Add and Search Words Data Structure/sol.py





import os
import json
from typing import Optional



class TrieNode:
    def __init__(self):
        self.children = {}
        self.endOfWord = False

class WordDictionary:

    def __init__(self):
        self.root = TrieNode()
        

    def addWord(self, word: str) -> None:
        curr = self.root 
        for c in word:
            if c not in curr.children:
                curr.children[c] = TrieNode()
            curr = curr.children[c]
        curr.endOfWord = True

    def searchFromNode(self, word: str, curr: Optional[TrieNode]) -> bool:
        for i in range(len(word)):
            c = word[i]
            # print(c, curr.children)
            if c in curr.children:
                curr = curr.children[c]
                continue
            
            if c == "." and len(curr.children) != 0:
                # print("len of childs ", len(curr.children), curr.children)
                valid = False
                for child in curr.children.values():
                    if self.searchFromNode(word[i + 1:], child):
                        return True
                return False
            else:
                # print("dfas",curr)
                return False
        if curr.endOfWord:
            return True
        return False


    def search(self, word: str) -> bool:
      
        return self.searchFromNode(word, self.root)
            
        


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)

def main():
    filename = "testcase.json"
    current_filename = os.path.abspath(__file__)
    # print("file:",__file__, "current_filename:", current_filename)
    current_dir = os.path.dirname(current_filename)
    # print("current_dir:", current_dir)
    path = os.path.join(current_dir, filename)
    jsonData = open(path).read()
    testcases = json.loads(jsonData) 
    # print(testcases)
    passNum = 0
    wordDictionary = WordDictionary()
    tests = testcases["tests"]
    for test in tests:
        input = test["input"]
        operations = input["operations"]
        arguments = input["arguments"]
        expectedOutput = test["expectedOutput"]
        failed = False 
        output = []
        for i in range(len(operations)):
            res = None
            if operations[i] == "WordDictionary":
                wordDictionary = WordDictionary()
            if operations[i] == "addWord":
                wordDictionary.addWord(arguments[i][0])
            elif operations[i] == "search":
                res = wordDictionary.search(arguments[i][0])
            
            output.append(res)
        if output != expectedOutput:
            failed = True 
            print(f"Test failed")
        else:
            print(f"Test passed")
            passNum += 1

        print("inputs :" )
        print(" operations :", operations)
        print(" arguments  :", arguments)
        print(" output     :", output)
        print(" expected   :", expectedOutput)
        print("----------------")
    print("Total tests: ", len(tests),", Passed: ", passNum, ", failed: ", len(tests) - passNum)


if __name__ == "__main__":
    main()